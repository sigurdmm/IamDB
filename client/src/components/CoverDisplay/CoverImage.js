import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import defaultImage from '../InformationSection/no-image-found.jpg';

import './CoverImage.less';

const getRating = (rating) => {
  if (typeof rating !== 'undefined') {
    return <em className='cover__overlay__rating'>{rating}/10</em>;
  }
  return (null);
};

const getTitleAndYear = (title, released) => {
  if (typeof released === 'undefined') {
    return <strong className='cover__overlay__title'>{title}</strong>;
  }
  const timestamp = new Date(parseInt(released, 10));
  return <strong className='cover__overlay__title'>{`${title} (${timestamp.getFullYear()})`}</strong>;
};

const CoverImage = ({
  thumbnail,
  title,
  rating,
  released,
  url,
}) => <div className={`cover ${!thumbnail ? 'cover--inverse' : ''}`}>
  <Link to={url}>
    <img className='cover__thumbnail' src={thumbnail || defaultImage}/>

    <div className='cover__overlay'>
      {getTitleAndYear(title, released)}
      {getRating(rating)}
    </div>
  </Link>
</div>;

CoverImage.propTypes = {
  thumbnail: PropTypes.string,
  title: PropTypes.string.isRequired,
  rating: PropTypes.number,
  url: PropTypes.string,
  released: PropTypes.any,
};

export default CoverImage;
