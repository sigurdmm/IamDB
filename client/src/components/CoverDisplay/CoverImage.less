@import '../../constants';

.cover {
  max-height: 200px;
  width: 120px;
  height: 400px;
  position: relative;
  display: inline-block;
  overflow: hidden;
  margin: 0;

  &__thumbnail {
    display: block;
    max-width: 120px;
    margin-left: auto;
    margin-right: auto;
    opacity: 0.3;

    position: absolute;

    top: 50%;
    left: 50%;

    min-height: 100%;
    min-width: auto;
    transform: translate(-50%, -50%);
  }

  img:hover {
    opacity: 1.0;
  }

  &__overlay {
    padding: 4px;
    font-weight: bold;
    color: @textColor;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    height: 100%;

    &__title {

    }

    &__rating {
    }
  }
}
