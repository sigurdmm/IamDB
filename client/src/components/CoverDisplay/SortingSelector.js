import PropTypes from 'prop-types';
import React from 'react';

const SortingSelector = ({
  sortingMethods,
  onSort,
}) => {
  // If onSort is false don't render SortingSelector
  // || typeof onSortDirection === 'undefined'
  if (typeof onSort === 'undefined') {
    return null;
  }

  return <div>
    <select name="select" onChange={event => onSort(event.target.value)}>
    {sortingMethods.map((method, i) => (
      <option
        key={i}
        value={method.value}
      >
        {method.label}
      </option>))}
    </select>
  </div>;
};

SortingSelector.propTypes = {
  onSort: PropTypes.func,
  sortingMethods: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired,
  })).isRequired,
};

export default SortingSelector;
