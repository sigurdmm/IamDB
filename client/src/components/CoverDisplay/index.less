@import '../../constants';

.coverdisplay {
  display: grid;
  background-color: @white;
  grid-gap: 1em;
  grid-template-columns: repeat(auto-fill, minmax(8em, 1fr));
  padding: 1em;
}
.coverdisplay > cover {
  margin-left: auto;
  margin-right: auto;
}

@media only screen and (max-width: 340px) {
  .coverdisplay {
    grid-template-columns: none;
    justify-content: center;
  }
}


