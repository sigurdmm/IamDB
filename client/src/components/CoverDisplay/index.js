import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import './index.less';
import CoverImage from './CoverImage';
import LoadingSpinner from './LoadingSpinner';

export default class CoverDisplay extends PureComponent {
  static propTypes = {
    media: PropTypes.array,
    pagination: PropTypes.element,
    hasSearched: PropTypes.bool.isRequired,
    url: PropTypes.string,
    error: PropTypes.string,
    loading: PropTypes.bool.isRequired,
  };

  render() {
    const {
      media,
      pagination,
      error,
      loading,
    } = this.props;
    // Don't render the CoverDisplay if users hasn't searched yet.
    if (!this.props.hasSearched) {
      return null;
    }
    // Render error message instead of an empty CoverDisplay
    // when search returns nothing
    if (error !== null) {
      return <p className='coverdisplay__error'>{error}</p>;
    }
    if (media.length === 0 && !loading) {
      return <p className='coverdisplay__error'>Nothing Found</p>;
    }
    if (loading) {
      return <LoadingSpinner/>;
    }

    return <>
      <div className='coverdisplay'>
        {media.map((cover, i) => <CoverImage
          thumbnail={cover.thumbnails ? cover.thumbnails.small : null}
          title={cover.name}
          rating={cover.rating}
          key={`cover-${i}`}
          url={`${this.props.url}${cover.id}`}
          released={cover.released}/>)}
      </div>
    { media.length > 0 && pagination }
    </>;
  }
}
