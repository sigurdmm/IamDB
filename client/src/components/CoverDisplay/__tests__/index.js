import { render } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';

import CoverDisplay from '../index';

const mockedMedia = [
  {
    id: '1',
    thumbnails: {
      small: 'https://example.com/test.jpg',
    },
  },
  {
    id: '2',
    thumbnails: {
      small: 'https://example.com/test2.png',
    },
  },
  {
    id: '3',
    thumbnails: {},
  },
];

describe('<CoverDisplay/>', () => {
  it('should render list properly', () => {
    const wrapper = render(<MemoryRouter>
      <CoverDisplay hasSearched={false} media={mockedMedia}/>
    </MemoryRouter>);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render error message properly', () => {
    const wrapper = render(<MemoryRouter>
      <CoverDisplay hasSearched={false} media={[]}/>
    </MemoryRouter>);

    expect(wrapper).toMatchSnapshot();
  });
});
