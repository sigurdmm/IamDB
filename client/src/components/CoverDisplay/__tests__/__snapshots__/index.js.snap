// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CoverDisplay/> should render error message properly 1`] = `null`;

exports[`<CoverDisplay/> should render list properly 1`] = `null`;
